doc/source/guide/in_depth/functional_guide.rst

```diff
@@ -0,0 +1,257 @@
+ * the field of scalars ``F`` are the real numbers.
+
+The first assumption is made in order to simplify the concept of *convex conjugate functional*, see ``conjugate_functional`` under :ref:`implementation` for more details, or the Wikipedia articles on `convex conjugates`_ and `Legendre transformations`_.
+
+The second assumption is made in order to guarantee that the optimization problem, in some way, makes sense.
+For example, if the field is not (totally) ordered (like for example the complex numbers) one cannot define optimization over it in a meaningful way.
+Or if the field does not have the least-upper-bound property, then attainment of an optimal solution cannot be guaranteed.
+See, for example, the Wikipedia articles on `field`_, `ordered field`_ and `least-upper-bound property`_.
+
+Note that it is possible to use the class without fulfilling these assumtptions, however in this case some of the mathematical results might not hold.
+
+.. _convex conjugate: https://en.wikipedia.org/wiki/Convex_conjugate
+.. _Legendre transformation: https://en.wikipedia.org/wiki/Legendre_transformation
+
+.. _field: https://en.wikipedia.org/wiki/Field_(mathematics)
+.. _ordered field: https://en.wikipedia.org/wiki/Ordered_field
+.. _least-upper-bound property: https://en.wikipedia.org/wiki/Least-upper-bound_property
+
+.. _implementation:
+
+Implementation of functionals
+=============================
+
+To define your own functional, you start by writing::
+
+    class MyFunctional(odl.solvers.Functional):
+        ...
+
+Since `Functional` is a subclass of `Operator`, it has the *abstract method* ``domain`` which need to be explicitly overridden by any subclass.
+
+``domain``: `Set`
+    The domain of this functional, i.e., the set of elements to which this functional can be applied.
+
+Moreover, there are several optional parameters associated with a functional. These are
+
+``linear``: `bool`, optional
+    If `True`, the functional is considered as linear.
+    Default: ``False``
+``smooth``: `bool`, optional
+    If `True`, assume that the functional is continuously differentiable.
+    Default: ``False``
+``concave``: `bool`, optional
+    If `True`, assume that the functional is concave.
+    Default: ``False``
+``convex``: `bool`, optional
+    If `True`, assume that the functional is convex.
+    Default: ``False``
+``grad_lipschitz``: `float`, optional
+    The Lipschitz constant of the gradient.
+    Default: infinity
+
+A functional also has a number of optional methods and properties associated with it.
+The default value of these are all to raise a `NotImplemetedError`.
+The properties are:
+
+ * ``gradient``. This returns the gradient operator of the functional, i.e., the operator that corresponds to the mapping
+
+   .. math::
+
+      x \to \nabla f(x),
+
+   where :math:`\nabla f(x)` is the element used to evaluate the (directional) derivative in a direction :math:`d` by :math:`\langle \nabla f(x), d \rangle`.
```

Comment:
Hm, maybe the connection to `Operator.derivative` could be clearer. Having the link to the method would be good, and one to our operator intro if it makes sense. I'd also write that the gradient at a point `f` is the space element representing the (Frechet) derivative at the same point:
```
.. math
   S'(f)(g) = \langle \nabla S(f), g \rangle \quad \text{for all } g \in X.
```